extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java

```diff
@@ -224,23 +286,62 @@ private static JsonObject getRolesJson(RoutingContext vertxContext, TenantConfig
             TenantConfigContext resolvedContext) {
-        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super AccessToken>>() {
-            @Override
-            public void accept(UniEmitter<? super AccessToken> uniEmitter) {
-                resolvedContext.auth.decodeToken((String) vertxContext.get("access_token"),
-                        new Handler<AsyncResult<AccessToken>>() {
-                            @Override
-                            public void handle(AsyncResult<AccessToken> event) {
-                                if (event.failed()) {
-                                    uniEmitter.fail(new AuthenticationFailedException(event.cause()));
+        if (request.getToken() instanceof IdTokenCredential
+                && (resolvedContext.oidcConfig.authentication.verifyAccessToken
+                        || resolvedContext.oidcConfig.roles.source.orElse(null) == Source.accesstoken)) {
+            final String codeAccessToken = (String) vertxContext.get("access_token");
+            if (OidcUtils.isOpaqueToken(codeAccessToken)) {
+                // remote introspection is required, a blocking call
+                if (BlockingOperationControl.isBlockingAllowed()) {
+                    return Uni.createFrom().emitter(
+                            new Consumer<UniEmitter<? super AccessToken>>() {
+                                @Override
+                                public void accept(UniEmitter<? super AccessToken> uniEmitter) {
+                                    verifyCodeFlowAccessToken(uniEmitter, resolvedContext, codeAccessToken);
                                 }
-                                uniEmitter.complete(event.result());
-                            }
-                        });
+                            });
+                } else {
+                    return Uni.createFrom().emitter(
+                            new Consumer<UniEmitter<? super AccessToken>>() {
+                                @Override
+                                public void accept(UniEmitter<? super AccessToken> uniEmitter) {
+                                    tenantResolver.getBlockingExecutor().execute(new Runnable() {
+                                        @Override
+                                        public void run() {
+                                            verifyCodeFlowAccessToken(uniEmitter, resolvedContext, codeAccessToken);
+                                        }
+                                    });
+                                }
+                            });
+                }
+            } else {
+                return Uni.createFrom().emitter(new Consumer<UniEmitter<? super AccessToken>>() {
+                    @Override
+                    public void accept(UniEmitter<? super AccessToken> uniEmitter) {
+                        verifyCodeFlowAccessToken(uniEmitter, resolvedContext, codeAccessToken);
```

Comment:
I don't get why sometimes it must be wrapped on an executor and sometimes not.

Reply:
@cescoffier Please see above, [this](https://github.com/quarkusio/quarkus/pull/13257/files#r522163580) and [this](https://github.com/quarkusio/quarkus/pull/13257/files#r522168791) comments